tests/lib/devicetree/api/src/main.c

```diff
@@ -2315,6 +2315,80 @@ static void test_string_token(void)
 	}
 }
 
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT vnd_string_array_token
+static void test_string_idx_token(void)
+{
+	enum token_string_idx {
+		/* Tokens */
+		token_first_idx_zero,
+		token_first_idx_one,
+		token_first_idx_two,
+		token_second_idx_zero,
+		token_second_idx_one,
+		token_second_idx_two,
+		token_second_idx_three,
+		/* Upper tokens */
+		TOKEN_FIRST_IDX_ZERO,
+		TOKEN_FIRST_IDX_ONE,
+		TOKEN_FIRST_IDX_TWO,
+		TOKEN_SECOND_IDX_ZERO,
+		TOKEN_SECOND_IDX_ONE,
+		TOKEN_SECOND_IDX_TWO,
+		TOKEN_SECOND_IDX_THREE
+	};
+
+	/* Test direct idx access */
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 0), token_first_idx_zero, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 1), token_first_idx_one, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 2), token_first_idx_two, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 0), token_second_idx_zero, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 1), token_second_idx_one, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 2), token_second_idx_two, "");
+	zassert_equal(DT_STRING_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 3), token_second_idx_three, "");
+
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 0), TOKEN_FIRST_IDX_ZERO, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 1), TOKEN_FIRST_IDX_ONE, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_first), val, 2), TOKEN_FIRST_IDX_TWO, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 0), TOKEN_SECOND_IDX_ZERO, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 1), TOKEN_SECOND_IDX_ONE, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 2), TOKEN_SECOND_IDX_TWO, "");
+	zassert_equal(DT_STRING_UPPER_TOKEN_BY_IDX(DT_NODELABEL(test_string_array_token_second), val, 3), TOKEN_SECOND_IDX_THREE, "");
+
+	/* Test instances */
+#define STRING_TOKEN_BY_IDX_VAR(node_id) _CONCAT(var_token_, node_id)
```

Comment:
I think this is quite a clever way to test instance-based APIs without relying on the order of instance numbers. Nice work; this trick hadn't occurred to me.